pinot-core/src/main/java/com/linkedin/pinot/core/segment/index/loader/LoaderUtils.java

```diff
@@ -90,4 +97,28 @@ public static void writeIndexToV3Format(SegmentDirectory.Writer segmentWriter, S
     }
     return stringList;
   }
+
+  /**
+   * Try to recover a segment from reload failure, no effect for normal segments.
+   * <p>Should be called before trying to load the segment.
+   */
+  public static void reloadFailureRecovery(@Nonnull File indexDir)
+      throws IOException {
+    File parentDir = indexDir.getParentFile();
+    File segmentTempDir = new File(parentDir, indexDir.getName() + CommonConstants.Segment.SEGMENT_TEMP_DIR_SUFFIX);
+    if (segmentTempDir.exists()) {
+      LOGGER.info("Trying to delete segment temporary directory: {}", segmentTempDir);
+      FileUtils.forceDelete(segmentTempDir);
+    }
+    File segmentBackupDir = new File(parentDir, indexDir.getName() + CommonConstants.Segment.SEGMENT_BACKUP_DIR_SUFFIX);
+    if (segmentBackupDir.exists()) {
+      LOGGER.info("Trying to recover index directory: {} from segment backup directory: {}", indexDir,
+          segmentBackupDir);
+      if (indexDir.exists()) {
+        FileUtils.forceDelete(indexDir);
+      }
+      Preconditions.checkState(segmentBackupDir.renameTo(indexDir),
```

Comment:
add a comment here that this step should be atomic, so it is ok if a failure happens during failure recovery. We will be left with the BACKUP directory, and can recover from that.